internal/service/cluster_mapping.go

```diff
@@ -0,0 +1,121 @@
+package service
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+
+	merrors "github.com/RedHatInsights/insights-operator-gathering-conditions-service/internal/errors"
+	"github.com/blang/semver/v4"
+	"github.com/rs/zerolog/log"
+)
+
+// ClusterMapping map OCP versions to remote configurations
+type ClusterMapping [][]string
+
+// IsValid check the list is in order (based on the versions), that the versions
+// can be parsed and that the remote configurations are accessible
+func (cm ClusterMapping) IsValid(remoteConfigurationPath string) bool {
```

Comment:
This might be just my ignorance (sorry about that). Would these checks be performed in the CI already? The mapping will be maintained manually. I think it would be best if the CI prevented devs from merging mappings that don't meet these checks in the first place. I cannot figure out if this MR or https://github.com/RedHatInsights/insights-operator-gathering-conditions-service/commit/cddf1e76944c266b55ccd0b988e56710c27f578e do this or not; I don't think they do.

Reply:
The CI involves ephemeral tests which deploys the service. The `isValid` function makes the service fail on startup if the cluster mapping is not valid. We can write a specific check for this on the CI, but tests should be enough. Maybe it would be faster to write a custom check to avoid waiting for ephemeral tests to run though

Reply:
PTAL @jholecek-rh : https://github.com/RedHatInsights/insights-operator-gathering-conditions-service/pull/266/commits/f8d53d6dcfa2d3929e3bd41ca4286a19cfea1b94 :)

Reply:
and the CI check fails because the conditions were not built into the image running the tests: https://github.com/RedHatInsights/insights-operator-gathering-conditions-service/actions/runs/10489571267/job/29054494304?pr=266#step:15:6
I'll fix it

Reply:
ok finally passing -> https://github.com/RedHatInsights/insights-operator-gathering-conditions-service/actions/runs/10490166296/job/29056308756?pr=266#step:15:1

Reply:
now we can add extra checks to the `isValid` function to check for the minumun version f.e.

Reply:
IIUC the ephemeral tests are not executed on every commit. I thought this check is cheap enough to be executed on every commit, giving the feedback sooner and cheaper than with ephemeral tests. https://github.com/RedHatInsights/insights-operator-gathering-conditions-service/commit/f8d53d6dcfa2d3929e3bd41ca4286a19cfea1b94 does the trick. Thanks!

Reply:
As for the minimum version test, yes, we can add it to the `isValid` function. It is a convenient place in the current code.

I am just not sure about the split of responsibilities between the service executable, service configuration/data, and tests.

* The minimum version supported by a mapping file is a property of the mapping file. The service executable doesn't need to care what the minimum version is to do its job. It doesn't have to be hardcoded in the service executable and it does not belong to service configuration. It is a property of service configuration.
* Similarly, the service executable doesn't need to require a sorted mapping file. It could easily sort it itself (it does anyway when it checks that it is sorted). This is a convention to make life easier for human operators editing the file. I see it as a responsibility of a linter-like test to make sure that the convention is followed.

Implementing these checks in the service executable extends its scope. The I-like-good-test-coverage me would then like to see tests for these extra features of the service executable, but it is like writing tests for tests.  If we skip these "tests for tests", we make reasoning about test coverage more difficult for newcomers (and future us).

P.S. Take this with a grain of salt. I am primarily a Python developer where separating "linter-like checks"  from production code wouldn't involve building a separate executable. :-)

Reply:
With regard to sorting I agree with Jan. I can't see anything wrong with unsorted map. The check seems to me unnecessary.

Reply:
Not what I meant. I was questioning what is the scope of the service and what is the scope of its tests. It is not a big deal. Let's leave it as it is.

@tremes I intentionally want the file to be sorted. It will help human operators interpret and edit the JSON document. I don't think you could assert the sorting using standard JSON parsers the data were stored in a map instead of a list. We were just discussing whether the check should be in the scope of the service or tests (like a linter for the JSON document).

The current implementation assumes that the list is ordered. In that case, it is desirable that the service validates that on startup, IMHO. I was suggesting that the implementation would not have to assume an ordered list.